### PR TITLE
Improve the type hints of synapse.api.errors.

### PR DESCRIPTION
**Pull request by clokep**
I started doing this in #7802 and realized it was bigger than I wanted so I split it out.

It adds more type hints to `synapse.api.errors`.

### CODE REVIEW

File: synapse/api/errors.py [hunk 1]
```diff
@@ -226,11 +232,11 @@ class InteractiveAuthIncompleteError(Exception):
     (This indicates we should return a 401 with 'result' as the body)
 
     Attributes:
-        result (dict): the server response to the request, which should be
+        result: the server response to the request, which should be
             passed back to the client
     """
 
-    def __init__(self, result):
+    def __init__(self, result: Dict):
```
Comment (erikjohnston) [inline synapse/api/errors.py:239]:
```suggestion
    def __init__(self, result: JsonDict):
```

or just `dict`

File: synapse/api/errors.py [hunk 2]
```diff
@@ -382,10 +389,10 @@ class LimitExceededError(SynapseError):
 
     def __init__(
         self,
-        code=429,
-        msg="Too Many Requests",
-        retry_after_ms=None,
-        errcode=Codes.LIMIT_EXCEEDED,
+        code: int = 429,
+        msg: str = "Too Many Requests",
+        retry_after_ms: int = None,
```
Comment (erikjohnston) [inline synapse/api/errors.py:394]:
```suggestion
        retry_after_ms: Optional[int] = None,
```

**Comment by clokep** [comment]
> Does this now pass `mypy synapse/api/errors.py`?

`synapse/api` is already included in mypy.

File: synapse/api/errors.py (hunk 1 above)
Comment (clokep) [reply synapse/api/errors.py:239]: Ugh, this makes a cycle. Good call though! 👍

**Comment by erikjohnston** [comment]
> > Does this now pass `mypy synapse/api/errors.py`?
> 
> `synapse/api` is already included in mypy.

I really expected that it would have picked up the `Optional` issue :confused:

**Comment by clokep** [comment]
> I really expected that it would have picked up the `Optional` issue 😕

I think mypy is "smart" in that case (when the default is `None`) and automatically adds `Optional`. Personally I find it confusing and like to be explicit.

**Pull request merged**